### PR TITLE
Update: object-curly-spacing spaceInEmptyObject option (fixes #7927)

### LINKED ISSUE #7927: New option in "object-curly-spacing" for empty objects
**Tell us about your environment**

* **ESLint Version:** v3.8.0
* **Node Version:** v6.7.0
* **npm Version:** v3.10.7

**What parser (default, Babel-ESLint, etc.) are you using?**

Default

**Please show your full configuration:**

```
{
  "rules": {
    "object-curly-spacing": ["error", "never"]
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
var a = { }
```

**What did you expect to happen?**

I expected that the rules should complain about the stray space between the curly braces { }

I also think it would be a good idea to add a separate config for empty objects, since I think that people who have this rule set to "always" still doesn't want any whitespace inside empty objects.

**What actually happened? Please include the actual, raw output from ESLint.**

No error was raised.

(Note: Re-opening https://github.com/eslint/eslint/issues/7380 for discussion, as suggested)

### PR DESCRIPTION
**Pull request by pmcelhaney**
**What is the purpose of this pull request? (put an "X" next to item)**

[] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

https://github.com/eslint/eslint/issues/7927

**What changes did you make? (Give an overview)**

Added a new option, "spaceInEmptyObject". It can be "never" or "always". The default is to ignore empty objects, as the rule does now. 

**Is there anything you'd like reviewers to focus on?**

No.

### CODE REVIEW

**Comment by platinumazure** [comment]
@eslint/eslint-team Issue #7927 was originally accepted as a bug, where the resolution seemed to be that `object-curly-spacing: ["error", "always"]` should enforce `{ }` rather than `{}`. This PR is instead an enhancement to add an optional option that can enforce one way or the other, and the absence of the option will allow both styles for empty objects. I think we need to evaluate this enhancement request separately from the original bug, but bearing in mind that it should be a good option for fixing that bug.

I'll champion this. We need three :+1:s from other team members. Let's do this.

File: lib/rules/object-curly-spacing.js [hunk 1]
```diff
@@ -32,6 +32,9 @@ module.exports = {
                     },
                     objectsInObjects: {
                         type: "boolean"
+                    },
+                    spaceInEmptyObject: {
+                        type: "string"
```
Comment (platinumazure) [inline lib/rules/object-curly-spacing.js]: I think this should be an `enum` so we can enforce always/never specifically. (See line 25 for another example)

File: tests/lib/rules/object-curly-spacing.js [hunk 2]
```diff
@@ -58,6 +58,9 @@ ruleTester.run("object-curly-spacing", rule, {
 
         // always - empty object
         { code: "var foo = {};", options: ["always"] },
+        { code: "var foo = {};", options: ["always", { spaceInEmptyObject: "never" }] },
+        { code: "var foo = { };", options: ["always", { spaceInEmptyObject: "always" }] },
```
Comment (platinumazure) [inline tests/lib/rules/object-curly-spacing.js]: Could you add some tests for line breaks? (I assume those should always pass and should be handled by the `object-curly-newline` rule.)

```js
var foo = {
};
// Should pass(?) for all combinations of first option + spaceInEmptyObject option
```

File: docs/rules/object-curly-spacing.md [hunk 3]
```diff
@@ -37,6 +37,8 @@ Object option:
 * `"arraysInObjects": false` disallows spacing inside of braces of objects beginning and/or ending with an array element (applies when the first option is set to `always`)
 * `"objectsInObjects": true` requires spacing inside of braces of objects beginning and/or ending with an object element (applies when the first option is set to `never`)
 * `"objectsInObjects": false` disallows spacing inside of braces of objects beginning and/or ending with an object element (applies when the first option is set to `always`)
+* `"spaceInEmptyObject": "always"`: requires a space inside an empty object literal (`{ }` is valid, `{}` is invalid)
+* `"spaceInEmptyObject": "never"`: disallows a space inside an empty object literal (`{}` is valid, `{ }` is invalid)
```
Comment (not-an-aardvark) [inline docs/rules/object-curly-spacing.md]: For consistency with the other options, can we change the name and schema to `emptyObjects: true` or `emptyObjects: false`?

File: tests/lib/rules/object-curly-spacing.js [hunk 4]
```diff
@@ -120,6 +123,8 @@ ruleTester.run("object-curly-spacing", rule, {
 
         // never - empty object
         { code: "var foo = {};", options: ["never"] },
+        { code: "var foo = {};", options: ["never", { spaceInEmptyObject: "never" }] },
+        { code: "var foo = { };", options: ["never", { spaceInEmptyObject: "always" }] },
```
Comment (not-an-aardvark) [inline tests/lib/rules/object-curly-spacing.js]: Can you add a test where an empty object contains a comment?

```js
var x = { /* no properties */ }
```

File: docs/rules/object-curly-spacing.md (hunk 3 above)
Comment (pmcelhaney) [reply docs/rules/object-curly-spacing.md]: I'm going for consistency with the main option.

**Comment by pmcelhaney** [comment]
@platinumazure @not-an-aardvark Thanks for the quick feedback! I've added tests for comments, new lines and a mix of various white space. Also changed the option to an emum.

File: docs/rules/object-curly-spacing.md (hunk 3 above)
Comment (not-an-aardvark) [reply docs/rules/object-curly-spacing.md]: Thanks for explaining. It looks like there is already some inconsistency here (between `always`/`never` for the main option, and `true`/`false` for the object options). Having said that, I still think it would be good to maintain consistency *among* the object options, so that we don't end up with one object option that arbitrarily has different semantics from the other options. To that end, I still think it would be better to use `true`/`false` here, and to give it a plural name without the word `space` (such as `emptyObjects`).

Comment (platinumazure) [reply docs/rules/object-curly-spacing.md]: I can see the argument for consistency, but personally I think always/never is more clear. True/false could be read as either, enforce space/enforce no space (i.e., equivalent to always/never), or it could be read as enforce in this case/don't enforce in this case (i.e., false = ignore, and true would mean always or never specifically).

I'm willing to bow towards consistency, but I think we should seriously reconsider the use of boolean options when the false could be read to mean either enforcing a lack of something, or not enforcing something.

Comment (not-an-aardvark) [reply docs/rules/object-curly-spacing.md]: I would also be fine with changing all of the options to accept `always`/`never` (and deprecating the boolean versions). I just don't want to enforce one convention for one option, and a different convention for another option.

File: tests/lib/rules/object-curly-spacing.js [hunk 5]
```diff
@@ -120,6 +125,12 @@ ruleTester.run("object-curly-spacing", rule, {
...
+        { code: "var foo = {/* this is a comment */};", options: ["never", { spaceInEmptyObject: "always" }] },
```
Comment (not-an-aardvark) [inline tests/lib/rules/object-curly-spacing.js]: I think this case will currently report an error with `spaceInEmptyObject: "never"`, but I don't think that's desirable -- the rule shouldn't forbid people from putting comments in objects. Instead, I would expect the rule to enforce or forbid spaces *around* the comment in this case, depending on either the main option or the `spaceInEmptyObject` option.

File: docs/rules/object-curly-spacing.md (hunk 3 above)
Comment (pmcelhaney) [reply docs/rules/object-curly-spacing.md]: Before making this change, I want to make sure we're clear.

These values will remain for backward compatibility. At least until the next major release, maybe longer.

> * `"objectsInObjects": true` requires spacing inside of braces of objects beginning and/or ending with an object element (applies when the first option is set to `never`)
 > * `"objectsInObjects": false` disallows spacing inside of braces of objects beginning and/or ending with an object element (applies when the first option is set to `always`)

These new values will be added. Note the big difference is "objectsInObjects" et. al. are now decoupled from the first option. 

 > * `"objectsInObjects": "always"` requires spacing inside of braces of objects beginning and/or ending with an object element 
 > * `"objectsInObjects": "never"` disallows spacing inside of braces of objects beginning and/or ending with an object element 

Should I bring this up in the issue thread? I'd like to have consensus before moving forward.

File: tests/lib/rules/object-curly-spacing.js (hunk 5 above)
Comment (pmcelhaney) [reply tests/lib/rules/object-curly-spacing.js]: Are you saying it *should* report an error? I don't think the ~~"spaceInEmptyObject~~ "emptyObjects" option should have any bearing on an object that contains anything other than whitespace. 

Instead, we need yet another option for "commentsInObjects".

File: docs/rules/object-curly-spacing.md (hunk 3 above)
Comment (not-an-aardvark) [reply docs/rules/object-curly-spacing.md]: If we're going to allow `always`/`never` as aliases for `true`/`false` for all the options, I think we might want to make that a separate issue and PR so that we don't block adding this option. My preference would be to use `true`/`false` with `emptyObjects` for now so that we can remain consistent, and then consider allowing `always`/`never`  for all the options separately.

File: tests/lib/rules/object-curly-spacing.js (hunk 5 above)
Comment (not-an-aardvark) [reply tests/lib/rules/object-curly-spacing.js]: I'm saying that the following code *should not* report an error, but it does report an error with the current implementation:

```js
/* eslint object-curly-spacing: [error, never, { spaceInEmptyObject: never }] */

var foo = {/* this is a comment */};
```

The issue with the current behavior is that `spaceInEmptyObject: "never"` is unsatisfiable for objects with a comment and no properties, because the rule still measures the offset between the `{` and the `}` rather than measuring the offset between the `{` and the start of the comment.

I don't feel strongly about how cases like the following should be treated:

```js
/* eslint object-curly-spacing: [error, always, { spaceInEmptyObject: never }] */

var foo = {/* this is a comment */};
```

I think I would want an error to be reported since there is no space between the `{` and the `/*`, and the object isn't "empty", but I'm open to being convinced otherwise.

**Comment by platinumazure** [comment]
When considering empty objects, should we define those by lack of tokens/comments between the curly braces, instead of lack of properties? I feel like that makes more sense conceptually. What do you guys think?

**Comment by not-an-aardvark** [comment]
Sounds good to me. However, that could cause this change to start reporting more errors if someone currently has code like this:

```js
/* eslint object-curly-spacing: [error, always] */

({/* comment */})
```

File: tests/lib/rules/object-curly-spacing.js (hunk 5 above)
Comment (pmcelhaney) [reply tests/lib/rules/object-curly-spacing.js]: Okay, yes, that should not report an error.

**Comment by platinumazure** [comment]
@eslint/eslint-team One more :+1: and we could accept this enhancement!

**Comment by btmills** [comment]
Just added my :+1: - this is now accepted!

**Comment by not-an-aardvark** [comment]
@pmcelhaney Friendly ping; do you think you will be able to fix the issue described in https://github.com/eslint/eslint/pull/9244#discussion_r137401272?

**Comment by pmcelhaney** [comment]
Yes. Sorry, didn’t realize / forgot there was work left to be done. Are you referring to the comment to which I replied, “Okay, yes, that should not report an error”?

**Comment by not-an-aardvark** [comment]
Yes, that's what I was referring to.

File: tests/lib/rules/object-curly-spacing.js [hunk 6]
```diff
@@ -811,6 +811,17 @@ ruleTester.run("object-curly-spacing", rule, {
                 }
             ]
         },
+        {
+            code: "var foo = { /* comment */ };",
+            output: "var foo = {};",
```
Comment (pmcelhaney) [inline tests/lib/rules/object-curly-spacing.js]: Oh no, that's no good. We don't want to delete the comment! I'll work on either removing the spaces and leaving the comment or skipping the fix in this case.

File: tests/lib/rules/object-curly-spacing.js [hunk 7]
```diff
@@ -58,6 +58,11 @@ ruleTester.run("object-curly-spacing", rule, {
 
         // always - empty object
         { code: "var foo = {};", options: ["always"] },
+        { code: "var foo = {};", options: ["always", { spaceInEmptyObject: "never" }] },
+        { code: "var foo = { };", options: ["always", { spaceInEmptyObject: "always" }] },
+        { code: "var foo = { /* this is a comment */ };", options: ["always", { spaceInEmptyObject: "always" }] },
+        { code: "var foo = {\n};", options: ["always", { spaceInEmptyObject: "always" }] },
```
[the remark this answers is not in the document]
Comment (pmcelhaney) [reply tests/lib/rules/object-curly-spacing.js]: Fixed so that a comment is not mistaken for a space.

**Comment by platinumazure** [comment]
Should the new option also handle space before the closing bracket? It looks like it might not right now based on my (admittedly very quick) skimming of the tests, but let me know if I'm mistaken. And if we aren't handling that space before the closing bracket-- should we?

**Comment by pmcelhaney** [comment]
In an empty object, the space preceding the closing bracket is the same as the space following the opening bracket. 

*Unless* one or more comment between the brackets break up the space(s). 

Having given it some more thought, it’s probably best not to trigger the rule at all if a comment is found within an empty object. That may come into conflict with a rule about spacing around comments.

**Comment by platinumazure** [comment]
> Having given it some more thought, it’s probably best not to trigger the rule at all if a comment is found within an empty object. That may come into conflict with a rule about spacing around comments.

This sounds like a sensible direction from this PR, and if we decide it makes sense to enforce in this rule, we can always change it later. :+1:

**Comment by platinumazure** [comment]
Is this ready for review from not-an-aardvark again, or do you still need to address some of his comments? (Sorry, it's a bit hard to tell in this interface with all of the comments he raised.)

**Comment by pmcelhaney** [comment]
Okay, I need to make some changes so that an empty object that contains a comment never triggers this rule. Will comment here when I'm done.

**Comment by pmcelhaney** [comment]
I'm going to close this PR. I think we should set it aside for now. The difference between the semantics of `true`/`false` and `"always"`/`"never"` is too confusing. 

I need to think about it some more, but I'm leaning toward approaching the change in three stages. 

1. Add tests for `{ }` to make sure the behavior for existing configs is not impacted.
2. Add `"always"`/`"never"` options for `objectsInObjects` and `arraysInObjects` and deprecate `true`/`false`. 
3. Apply this change, adding `spaceInEmptyObjects` with support for `"always"/"never"` only.
